pkg/reconciler/apis/apiexport/apiexport_controller.go

```diff
@@ -0,0 +1,489 @@
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+
+	old := obj
+	obj = obj.DeepCopy()
+
+	var errs []error
+	if err := c.reconcile(ctx, obj); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Regardless of whether reconcile returned an error or not, always try to patch status if needed. Return the
+	// reconciliation error at the end.
+
+	// If the object being reconciled changed as a result, update it.
+	if err := c.patchIfNeeded(ctx, old, obj); err != nil {
+		errs = append(errs, err)
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (c *controller) patchIfNeeded(ctx context.Context, old, obj *apisv1alpha1.APIExport) error {
+	specChanged := !equality.Semantic.DeepEqual(old.Spec, obj.Spec)
+	statusChanged := !equality.Semantic.DeepEqual(old.Status, obj.Status)
+
+	if specChanged && statusChanged {
+		klog.Fatalf("Programmer error: spec and status changed in same reconcile iteration")
+	}
+
+	if !specChanged && !statusChanged {
+		return nil
+	}
+
+	apiExportForPatch := func(apiExport *apisv1alpha1.APIExport) apisv1alpha1.APIExport {
+		var ret apisv1alpha1.APIExport
+		if specChanged {
+			ret.Spec = apiExport.Spec
+		} else {
+			ret.Status = apiExport.Status
+		}
+		return ret
+	}
+
+	clusterName := logicalcluster.From(old)
+	name := old.Name
+
+	oldData, err := json.Marshal(apiExportForPatch(old))
+	if err != nil {
+		return fmt.Errorf("failed to Marshal old data for APIExport %s|%s: %w", clusterName, name, err)
+	}
+
+	newForPatch := apiExportForPatch(obj)
+	newForPatch.ObjectMeta = metav1.ObjectMeta{
+		UID:             old.UID,
+		ResourceVersion: old.ResourceVersion,
+	} // to ensure they appear in the patch as preconditions
+	newData, err := json.Marshal(newForPatch)
+	if err != nil {
+		return fmt.Errorf("failed to Marshal new data for APIExport %s|%s: %w", clusterName, name, err)
+	}
+
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return fmt.Errorf("failed to create patch for APIExport %s|%s: %w", clusterName, name, err)
+	}
+
+	var subresources []string
+	if statusChanged {
+		subresources = []string{"status"}
+	}
+
+	_, err = c.kcpClusterClient.Cluster(clusterName).ApisV1alpha1().APIExports().Patch(ctx, obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, subresources...)
+	return err
+}
+
+func (c *controller) readThroughGetSecret(ctx context.Context, clusterName logicalcluster.LogicalCluster, ns, name string) (*corev1.Secret, error) {
+	secret, err := c.secretLister.Secrets(ns).Get(clusters.ToClusterAwareKey(clusterName, name))
+	if err == nil {
+		return secret, nil
+	}
+
+	// In case the lister is slow to catch up, try a live read
+	secret, err = c.kubeClusterClient.Cluster(clusterName).CoreV1().Secrets(ns).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func (c *controller) reconcile(ctx context.Context, apiExport *apisv1alpha1.APIExport) error {
```

Comment:
I liked the separation into controller and reconciler we used before. It made it clear how to unit test a reconciler and separated business logic nicely from plumbing. This pattern here in the file will be hard to maintain and will quickly diverge. With the separation you know what is standard plumbing and what is custom business logic.

Reply:
Fixing